### PR TITLE
Add Backblaze B2 objectstore backend

### PR DESCRIPTION
**Pull request by bcarrella**
I've just added [Backblaze B2](https://www.backblaze.com/b2/cloud-storage.html) Primary Object Storage support to Nextcloud

This is an example of config to make it works.
I've also added composer dependency to [B2 SDK](https://github.com/bcarrella/backblaze-b2)  in 3rdparty submodules.
Requires https://github.com/nextcloud/3rdparty/pull/269

```
<?php
$CONFIG = array (
  'objectstore' => 
  array (
    'class' => '\\OC\\Files\\ObjectStore\\B2',
    'arguments' => 
    array (
      'bucket_name' => 'YOUR-BUCKET-NAME',
      'autocreate' => true,
      'account_id' => 'YOUR-ACCOUNT-ID',
      'key_id' => 'YOUR-KEY-ID',
      'application_key' => 'YOUR-APPLICATION-KEY',
      'cacheParentDir' => '/path/to/your/writable/cache/directory',
    ),
  ),
  'datadirectory' => '/path/to/your/nextcloud/datadirectory',
);
```

### CODE REVIEW

File: .htaccess [hunk 1]
```diff
@@ -65,3 +65,7 @@ Options -Indexes
 <IfModule pagespeed_module>
   ModPagespeed Off
 </IfModule>
+#### DO NOT CHANGE ANYTHING ABOVE THIS LINE ####
+
+ErrorDocument 403 //
```
Comment (kesselb) [inline .htaccess]: Please revert this change

File: lib/private/Files/ObjectStore/B2.php [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+            $this->autoCreate = $parameters['autocreate'];
+        }
+    }
+
+    /**
+     * @return Client
+     */
+    private function getClient()
+    {
+        if (!$this->client) {
```
Comment (kesselb) [inline lib/private/Files/ObjectStore/B2.php]:
```suggestion
        if ($this->client === null) {
```

Is more strict

File: lib/private/Files/ObjectStore/B2.php [hunk 3]
```diff
@@ -0,0 +1,163 @@
+<?php
```
Comment (kesselb) [inline lib/private/Files/ObjectStore/B2.php]: Please use strict types for new files
```
<?php
declare(strict_types=1);
```

File: lib/private/Files/ObjectStore/B2.php [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    /**
+     * @return Client
+     */
+    private function getClient()
+    {
+        if (!$this->client) {
+            $this->client = new Client($this->accountId, 
+            [
```
Comment (kesselb) [inline lib/private/Files/ObjectStore/B2.php]: indentation looks off

```
$this->client = new Client(
	$this->accountId, 
	[
		'keyId' => $this->keyId, // optional if you want to use master key (account Id)
		'applicationKey' => $this->applicationKey,
		'version' => 2, // By default will use version 1
	],
	[
		'largeFileLimit' => (4*1024*1024*1024), // Lower limit for using large files upload support. Default: 3GB
		'cacheParentDir' => \OC::$server->getConfig()->getSystemValue('datadirectory'),
	]
);
```

File: lib/private/Files/ObjectStore/B2.php [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+                'keyId' => $this->keyId, // optional if you want to use master key (account Id)
+                'applicationKey' => $this->applicationKey,
+                'version' => 2, // By default will use version 1
+            ],
+            [
+                'largeFileLimit' => (4*1024*1024*1024), // Lower limit for using large files upload support. Default: 3GB
+                'cacheParentDir' => \OC::$server->getConfig()->getSystemValue('datadirectory'),
```
Comment (kesselb) [inline lib/private/Files/ObjectStore/B2.php]: I'm note sure if datadirectory is the right choice for this. You might inject IConfig by DependencyInjection and use it here (only if possible)

File: lib/private/Files/ObjectStore/B2.php [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+                }
+
+                if (!$bucketExists) {
+                    $this->bucket = $this->client->createBucket([
+                        'BucketName' => $this->bucketName,
+                        'BucketType' => Bucket::TYPE_PRIVATE,
+                        'KeepLastVersionOnly' => true
+                    ]);
+
+                    $buckets = $this->client->listBuckets(true);
```
Comment (kesselb) [inline lib/private/Files/ObjectStore/B2.php]: the result $buckets is not used. why do you call listBuckets here?

File: lib/private/Files/ObjectStore/B2.php [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    /** @var string */
+    private $keyId;
+    /** @var string */
+    private $applicationKey;
+    /** @var Client|null */
+    private $client = null;
+    /** @var bool  */
+    private $autoCreate = false;
+    /** @var Bucket|null */
+    private $bucket = null;
```
Comment (kesselb) [inline lib/private/Files/ObjectStore/B2.php]: bucket is never used (only if autoCreate is true and bucket does not exist). A) always provide a value for $bucket or B) dont use a class variable here

File: lib/private/Files/ObjectStore/B2.php [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    /**
+     * @param string $urn the unified resource name used to identify the object
+     * @param resource $stream stream with the data to write
+     * @throws \Exception when something goes wrong, message will be logged
+     */
+    public function writeObject($urn, $stream)
+    {
+        $file = $this->getClient()->upload([
```
Comment (kesselb) [inline lib/private/Files/ObjectStore/B2.php]: $file is not used

File: lib/private/Files/ObjectStore/B2.php [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    /**
+     * @param string $urn the unified resource name used to identify the object
+     * @return void
+     * @throws \Exception when something goes wrong, message will be logged
+     */
+    public function deleteObject($urn)
+    {
+        $fileDelete = $this->client->deleteFile([
```
Comment (kesselb) [inline lib/private/Files/ObjectStore/B2.php]: $fileDelete is not used

**Comment by kesselb** [comment]
Thank you for this contribution :tada:

**Comment by kesselb** [comment]
The 3rdparty update looks wrong to me. I guess you might fork http://github.com/nextcloud/3rdparty/, checkout and open a pull request with your dependency.

I think there is a problem with `obregonco/backblaze-b2` (and your fork `bcarrella/backblaze-b2`):

```
"illuminate/cache": ">=5.1.0",
"illuminate/container": ">=5.1.0",
"illuminate/filesystem": ">=5.1.0"
```

If we add these packages to nextcloud we add at lot of additional dependencies. Lets ask @nextcloud/server-triage about that ;)

**Comment by nickvergessen** [comment]
Can't external storages be done with an app nowerdays? I think it would be much better suited.

cc @icewind1991 @rullzer

**Comment by bcarrella** [comment]
> Can't external storages be done with an app nowerdays? I think it would be much better suited.
> cc @icewind1991 @rullzer

@nickvergessen this is not true, because as stated at:
[Configuring Object Storage as Primary Storage](https://docs.nextcloud.com//server/14/admin_manual/configuration_files/primary_storage.html)

> Contrary to using an object store as external storage, when an object store is used as primary storage, no metadata (names, directory structures, etc) is stored in the object store. The metadata is only stored in the database and the object store only holds the file content by unique identifier.
> 
> Because of this primary object stores usually perform better than when using the same object store as external storage but it restricts being able to access the files from outside of Nextcloud.

**Comment by MorrisJobke** [comment]
> Can't external storages be done with an app nowerdays? I think it would be much better suited.

Only if they are external storages and not the primary storage.

File: .htaccess (hunk 1 above)
Comment (bcarrella) [reply .htaccess]: Fixed in commit e45de39ca7e90a099b3e7eddc5952bab3f4cdf44

File: lib/private/Files/ObjectStore/B2.php (hunk 2 above)
Comment (bcarrella) [reply lib/private/Files/ObjectStore/B2.php]: Fixed in commit e45de39ca7e90a099b3e7eddc5952bab3f4cdf44

File: lib/private/Files/ObjectStore/B2.php (hunk 3 above)
Comment (bcarrella) [reply lib/private/Files/ObjectStore/B2.php]: Fixed in commit e45de39ca7e90a099b3e7eddc5952bab3f4cdf44

File: lib/private/Files/ObjectStore/B2.php (hunk 4 above)
Comment (bcarrella) [reply lib/private/Files/ObjectStore/B2.php]: Fixed in commit e45de39ca7e90a099b3e7eddc5952bab3f4cdf44

File: lib/private/Files/ObjectStore/B2.php (hunk 5 above)
Comment (bcarrella) [reply lib/private/Files/ObjectStore/B2.php]: In commit e45de39ca7e90a099b3e7eddc5952bab3f4cdf44 I made 'cacheParentDir' a config parameter to let user choose where to put cache files

File: lib/private/Files/ObjectStore/B2.php (hunk 6 above)
Comment (bcarrella) [reply lib/private/Files/ObjectStore/B2.php]: Fixed in commit e45de39ca7e90a099b3e7eddc5952bab3f4cdf44, I need to call listBuckets to invalidate cache contents. Client SDK doesn't have public method to invalidate cache. I'll write a new one ASAP.

File: lib/private/Files/ObjectStore/B2.php (hunk 7 above)
Comment (bcarrella) [reply lib/private/Files/ObjectStore/B2.php]: Fixed in commit e45de39ca7e90a099b3e7eddc5952bab3f4cdf44, I removed unused class variable

File: lib/private/Files/ObjectStore/B2.php (hunk 8 above)
Comment (bcarrella) [reply lib/private/Files/ObjectStore/B2.php]: Fixed in commit e45de39ca7e90a099b3e7eddc5952bab3f4cdf44

File: lib/private/Files/ObjectStore/B2.php (hunk 9 above)
Comment (bcarrella) [reply lib/private/Files/ObjectStore/B2.php]: Fixed in commit e45de39ca7e90a099b3e7eddc5952bab3f4cdf44

**Comment by bcarrella** [comment]
> The 3rdparty update looks wrong to me. I guess you might fork [http://github.com/nextcloud/3rdparty/](https://github.com/nextcloud/3rdparty/), checkout and open a pull request with your dependency.
> 
> I think there is a problem with `obregonco/backblaze-b2` (and your fork `bcarrella/backblaze-b2`):
> 
> ```
> "illuminate/cache": ">=5.1.0",
> "illuminate/container": ">=5.1.0",
> "illuminate/filesystem": ">=5.1.0"
> ```
> 
> If we add these packages to nextcloud we add at lot of additional dependencies. Lets ask @nextcloud/server-triage about that ;)

I've opened a PR for 3rdparty related to this enhancement.
To resolve problem I linked 3rdparty submodule to my fork bcarrella/3rdparty.

**Comment by bcarrella** [comment]
> The 3rdparty update looks wrong to me. I guess you might fork http://github.com/nextcloud/3rdparty/, checkout and open a pull request with your dependency.
> I think there is a problem with obregonco/backblaze-b2 (and your fork bcarrella/backblaze-b2):
> "illuminate/cache": ">=5.1.0",
> "illuminate/container": ">=5.1.0",
> "illuminate/filesystem": ">=5.1.0"
> 
> If we add these packages to nextcloud we add at lot of additional dependencies. Lets ask @nextcloud/server-triage about that ;)
> 
> I've opened a PR for 3rdparty related to this enhancement.
> To resolve problem I linked 3rdparty submodule to my fork bcarrella/3rdparty.

Please let me know if this is correct.
@nextcloud/server-triage @kesselb

**Comment by rullzer** [comment]
I'm unsure if this is the right approach. Adding this to server means we have to support it from now on. And it is not something any of the main engineers will often use.

@icewind1991 could we somehow make primary storage also available via apps?

**Comment by icewind1991** [comment]
> could we somehow make primary storage also available via apps?

The main problem is that primary storage needs to be available during installation, if we have a way to configure apps to be available during installation (and maintenance mode, while upgrading, etc) then it can work

**Comment by ChristophWurst** [comment]
I'm closing this PR for now. As outline by the others we have to

* Make it possible to always-enable certain apps
* Make it possible to have primary storage in apps
* Add the new storage provider

Doing it in these steps will make most sense.
